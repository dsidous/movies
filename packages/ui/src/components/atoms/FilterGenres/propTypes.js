import { array, func } from 'prop-types';

export const propTypes = {
  genres: array.isRequired,
  onChange: func.isRequired,
};

export const defaultProps = {};
