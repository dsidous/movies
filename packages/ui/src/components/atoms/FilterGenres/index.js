import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import { propTypes } from './propTypes';

const FilterGenres = ({ genres, onChange }) => {
  const [selectedGenres, setSelectedGenres] = useState({
    options: [],
    selectValue: [''],
  });

  useEffect(() => {
    const options = genres.map(genre => ({
      label: genre.name,
      value: String(genre.id),
    }));
    setSelectedGenres(s => ({ ...s, options }));
  }, [genres]);

  const handleChange = event => {
    setSelectedGenres({
      ...selectedGenres,
      selectValue: event.target.value,
    });
  };

  const handleClose = () => {
    onChange(selectedGenres.selectValue);
  };

  return (
    <FormControl className="filter-element-wrapper" variant="outlined">
      <InputLabel htmlFor="genres">Genres</InputLabel>
      <Select
        multiple
        value={selectedGenres.selectValue}
        onChange={handleChange}
        onClose={handleClose}
        input={<OutlinedInput id="genres" labelWidth={50} />}
        renderValue={selected => {
          if (selected.length === 1) {
            return <span>Select genres...</span>;
          }

          const list = selected
            .filter(select => select !== '')
            .map(
              select =>
                selectedGenres.options.find(option => option.value === select)
                  .label,
            );

          return list.join(', ');
        }}
      >
        <MenuItem value="" disabled>
          Select genres...
        </MenuItem>
        {selectedGenres.options.map(option => (
          <MenuItem key={option.label} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

FilterGenres.propTypes = propTypes;

export default FilterGenres;
