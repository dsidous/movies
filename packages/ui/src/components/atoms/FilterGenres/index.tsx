/** @jsxFrag React.Fragment */
import React, { useState, useEffect } from 'react';

import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import OutlinedInput from '@material-ui/core/OutlinedInput';

import { propTypes } from './propTypes';

import { Genres } from '../../../types/movie';

type SelectValue = string;

type Option = {
  label: string;
  value: string;
};

interface Props {
  genres: Genres[];
  onChange: (arg0: SelectValue) => void;
}

interface Selected {
  options: Option[] | null;
  selectValue: SelectValue;
}

const FilterGenres: React.FC<Props> = ({ genres, onChange }: Props) => {
  const [selectedGenres, setSelectedGenres] = useState<Selected>({
    options: null,
    selectValue: '',
  });

  useEffect(() => {
    const options = genres.map(genre => ({
      label: genre.name,
      value: String(genre.id),
    }));
    setSelectedGenres(s => ({ ...s, options }));
  }, [genres]);

  const handleChange = (event: React.ChangeEvent<HTMLButtonElement>) => {
    setSelectedGenres({
      ...selectedGenres,
      selectValue: event.target.value,
    });
  };

  const handleClose = () => {
    onChange(selectedGenres.selectValue);
  };

  return (
    <>
      <InputLabel htmlFor="genres">Genres</InputLabel>
      <Select
        multiple
        value={selectedGenres.selectValue}
        onChange={handleChange}
        onClose={handleClose}
        input={<OutlinedInput id="genres" labelWidth={50} />}
        renderValue={(selected: []) => {
          if (!selected.length) {
            return <span>Select genres...</span>;
          }

          const list = selected
            .filter(select => select !== '')
            .map(select => {
              if (selectedGenres.options && selectedGenres.options.length > 0) {
                return selectedGenres?.options?.find(
                  (option: Option) => option.value === select,
                ).label;
              }

              return null;
            });

          return list.join(', ');
        }}
      >
        <MenuItem value="" disabled>
          Select genres...
        </MenuItem>
        {selectedGenres.options?.map(option => (
          <MenuItem key={option.label} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </>
  );
};

FilterGenres.propTypes = propTypes;

export default FilterGenres;
