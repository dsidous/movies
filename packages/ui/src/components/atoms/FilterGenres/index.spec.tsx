import React from 'react';
import { shallow } from 'enzyme';
import FilterGenres from '.';

const mockProps = {
  genres: [
    {
      id: 1,
      name: 'genre',
    },
  ],
  value: 'genre',
  onChange: jest.fn(),
};

describe('Atoms/FilterGenres', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<FilterGenres {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });
});
