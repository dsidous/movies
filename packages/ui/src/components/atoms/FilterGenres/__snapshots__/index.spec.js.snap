// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/FilterGenres should render as expected 1`] = `ShallowWrapper {}`;
